server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -475,8 +480,19 @@ protected void doMerge(final ObjectMapper mergeWith) {
                 // no mapping, simply add it
                 merged = mergeWithMapper;
             } else {
-                // root mappers can only exist here for backcompat, and are merged in Mapping
-                merged = mergeIntoMapper.merge(mergeWithMapper);
+                if (mergeIntoMapper instanceof ObjectMapper) {
+                    ObjectMapper objectMapper = (ObjectMapper) mergeIntoMapper;
+                    merged = objectMapper.merge(mergeWithMapper, reason);
+                } else {
+                    assert mergeIntoMapper instanceof FieldMapper || mergeIntoMapper instanceof FieldAliasMapper;
+                    // If we're merging template mappings when creating an index, then a field definition always
+                    // replaces an existing one.
+                    if (reason == MergeReason.INDEX_TEMPLATE) {
```

Comment:
This is the most interesting part, we decide to overwrite leaf fields instead of merging them.